test/code/providers/config_scanner_test.cpp

```diff
@@ -0,0 +1,669 @@
+    CPPUNIT_TEST( testCnfScan_IncludeDirWithNonCnfFiles_IgnoreNonCnf );
+    CPPUNIT_TEST( testCnfScan_DirectivesFormCycles);
+    CPPUNIT_TEST( testCnfScan_CnfFileMissing_ReturnsFalse);
+    CPPUNIT_TEST( testCnfScan_InvalidDirectives_Ignored );
+
+    CPPUNIT_TEST_SUITE_END();
+
+private:
+
+public:
+    void setUp(void)
+    {
+    }
+
+    void tearDown(void)
+    {
+    }
+
+    void ValidateAuth(std::map< unsigned int, std::string >& expected, MySQL_Authentication& auth)
+    {
+        std::vector<unsigned int> authPorts;
+        auth.GetPortList( authPorts );
+        CPPUNIT_ASSERT_EQUAL(expected.size(), authPorts.size());
+        for (std::vector<unsigned int>::const_iterator it = authPorts.begin(); it != authPorts.end(); ++it)
+        {
+            MySQL_AuthenticationEntry authEntry;
+            auth.GetEntry(*it, authEntry);
+            CPPUNIT_ASSERT_EQUAL(expected[authEntry.port], authEntry.binding);
+        }
+    }
+
+    void CreateConfig(const std::wstring& binding, unsigned int port, bool withMySqld, std::vector< std::wstring >& conf)
+    {
+        std::wstringstream out;
+        conf.push_back(out.str());
+        conf.push_back(L"# A Comment line");
+        conf.push_back(L"[client]");
+        out.str(L"");
+        out << L"port     = " << port ;
+        conf.push_back(out.str());
+        conf.push_back(L"socket		= /var/run/mysqld/mysqld.sock");
+        conf.push_back(L"An irrelevant line without equalsTo");
+        conf.push_back(L"A line = with more than one =  ");
+        out.str(L"");
+        out << L"bind-address   = " << binding;
+        conf.push_back(out.str());
+        conf.push_back(L"");
+        if (withMySqld)
+        {
+            conf.push_back(L"[mysqld]");
+            conf.push_back(L"#");
+            conf.push_back(L"# * Basic Settings");
+            conf.push_back(L"#");
+            conf.push_back(L"user		= mysql");
+            conf.push_back(L"pid-file	= /var/run/mysqld/mysqld.pid");
+            conf.push_back(L"socket		= /var/run/mysqld/mysqld.sock");
+            conf.push_back(L"port = 3306");
+            conf.push_back(L"basedir		= /usr");
+            conf.push_back(L"datadir		= /var/lib/mysql");
+            out.str(L"");
+        }
+        conf.push_back(out.str());
+    }
+
+    void testCnfScan_NoDirectives()
+    {
+        MySQL_Authentication auth( SCXHandle<MySQL_AuthenticationDependencies>(new MySQL_ScannerTestAuthenticationDependencies()));
+        auth.Load();
+        std::map< std::wstring, std::vector< std::wstring > > fileContentsMap;
+        std::map< std::wstring, std::vector< std::wstring > > directoryMap;
+        std::map< unsigned int, std::string > portBindingMap;
+        std::vector < std::wstring > contents;
+
+        std::wstring cnfFileName = L"NoDirectives.cnf";
+        std::wstring bindAddress = L"172.22.13.1";
+        unsigned int port = 3505;
+        CreateConfig(bindAddress, port, true, contents);
```

Comment:
FWIW, I too did not like the tests. But I do not like the alternative either :cry: 

 I used an in-memory config file for the following reasons (in approximately decreasing order of priority):

1.  LICENSE and Copyright  -
 I do not know the ownership/copyright status of files like my.cnf from ubuntu/debian distributions of the package. Can I just copy it into our codebase (with minor modifications)?

2. Even when I am using real files and real directories, I will still need some sort of mapping between the test files and test directories and the entries in the file.
The tests are typically run from the folder $HOME/dev/bld-mysql/mysql/intermediate/Arch.
So, the test should introduce a mapping like one of the following

      /etc/mysql/mysql.d/mysqld.conf     ---> <TEST-DATA-DIR>/etc/mysql/mysql.d/mysqld.cnf 
    
     No matter how I compute the above mapping at the time of test execution, I will impose some restrictions on the test execution.  (testrunner must be run only from the directory that is at a fixed location relative to the test data - or copy the entire directory with test-data into the current folder).

3. To understand what a test does, I need to look at multiple files - test code, test data, each of the included files (or document it and hope that the documentation, test code and the test data are all in sync). 

4. A chance of false-positives is higher in case of test-data in files. (Test reports success but actually it is not testing the required data). This is again because the test logic is determined by multiple artifacts. 

I do think mocking is the right way. But I understand some of the above are my own personal preferences.  I just wanted to present the reasons I chose. Please let me know if you prefer the files.